packages/cdktf/lib/iterator.ts

```diff
@@ -0,0 +1,200 @@
+    // for example it could return a NumberListIterator whose iterator.key would be a number
+    return new ListIterator(list);
+  }
+
+  /**
+   * Creates a new iterator from a map
+   */
+  public static fromMap(
+    map:
+      | ComplexMap
+      | { [key: string]: any }
+      | { [key: string]: string }
+      | { [key: string]: number }
+      | { [key: string]: boolean }
+  ): MapIterator {
+    return new MapIterator(map);
+  }
+
+  /**
+   * @param attribute name of the property to retrieve
+   * @returns the given attribute of the current item iterated over as a string
+   */
+  getString(attribute: string): string {
+    return Token.asString(propertyAccess(this._getValue(), [attribute]));
+  }
+
+  /**
+   * @param attribute name of the property to retrieve
+   * @returns the given attribute of the current item iterated over as a number
+   */
+  getNumber(attribute: string): number {
+    return Token.asNumber(propertyAccess(this._getValue(), [attribute]));
+  }
+
+  /**
+   * @param attribute name of the property to retrieve
+   * @returns the given attribute of the current item iterated over as a (string) list
+   */
+  getList(attribute: string): string[] {
+    return Token.asList(propertyAccess(this._getValue(), [attribute]));
+  }
+
+  /**
+   * @param attribute name of the property to retrieve
+   * @returns the given attribute of the current item iterated over as a number list
+   */
+  getNumberList(attribute: string): number[] {
+    return Token.asNumberList(propertyAccess(this._getValue(), [attribute]));
+  }
+
+  /**
+   * @param attribute name of the property to retrieve
+   * @returns the given attribute of the current item iterated over as a map
+   */
+  getMap(attribute: string): { [key: string]: any } {
+    return Token.asAnyMap(propertyAccess(this._getValue(), [attribute]));
+  }
+
+  /**
+   * @param attribute name of the property to retrieve
+   * @returns the given attribute of the current item iterated over as a map of strings
+   */
+  getStringMap(attribute: string): { [key: string]: string } {
+    return Token.asStringMap(propertyAccess(this._getValue(), [attribute]));
+  }
+
+  /**
+   * @param attribute name of the property to retrieve
+   * @returns the given attribute of the current item iterated over as a map of numbers
+   */
+  getNumberMap(attribute: string): { [key: string]: number } {
+    return Token.asNumberMap(propertyAccess(this._getValue(), [attribute]));
+  }
+
+  /**
+   * @param attribute name of the property to retrieve
+   * @returns the given attribute of the current item iterated over as a map of booleans
+   */
+  getBooleanMap(attribute: string): { [key: string]: boolean } {
+    return Token.asBooleanMap(propertyAccess(this._getValue(), [attribute]));
+  }
+
+  /**
+   * @internal
+   */
+  protected _getValue(): any {
+    // uses a Lazy value to be able to access the current TerraformStack and pass it to ref()
+    return Lazy.anyValue(
```

Comment:
We haven't done a lot of testing with `Lazy`. One area that seems likely to have issues is with cross stack references. Should make sure to have some test cases with iterators across stacks.

Reply:
Yes, that sounds like a good test case to add

Reply:
You were right, cross stack references did not resolve properly with iterators (due to Lazy). I'm going to have a time-boxed look at it.

Reply:
ah wait, it does work 🎉 The test needed to call `stack.prepareStack()` for the cross stack references to be created.

Reply:
added a test ✅